src/lib.rs

```diff
@@ -153,6 +153,67 @@ fn number_weakly_connected_components(graph: &digraph::PyDiGraph) -> usize {
     algo::connected_components(graph)
 }
 
+/// Find the weakly connected components in a directed graph
+///
+/// :param PyDiGraph graph: The graph to find the weakly connected components
+///     in
+///
+/// :returns: A list of sets where each set it a weakly connected component of
+///     the graph
+/// :rtype: list
+#[pyfunction]
+#[text_signature = "(graph, /)"]
+pub fn weakly_connected_components(
+    graph: &digraph::PyDiGraph,
+) -> Vec<BTreeSet<usize>> {
+    let mut seen: HashSet<NodeIndex> = HashSet::new();
+    let mut out_vec: Vec<BTreeSet<usize>> = Vec::new();
+    for node in graph.graph.node_indices() {
+        if !seen.contains(&node) {
+            // BFS node generator
+            let mut component_set: BTreeSet<usize> = BTreeSet::new();
+            let mut bfs_seen: HashSet<NodeIndex> = HashSet::new();
+            let mut next_level: HashSet<NodeIndex> = HashSet::new();
+            next_level.insert(node);
+            while !next_level.is_empty() {
+                let this_level = next_level;
+                next_level = HashSet::new();
+                for bfs_node in this_level {
+                    if !bfs_seen.contains(&bfs_node) {
+                        component_set.insert(bfs_node.index());
+                        bfs_seen.insert(bfs_node);
+                        for neighbor in
+                            graph.graph.neighbors_undirected(bfs_node)
+                        {
+                            next_level.insert(neighbor);
+                        }
+                    }
+                }
+            }
+            out_vec.push(component_set);
+            seen.extend(bfs_seen);
+        }
+    }
+    out_vec
+}
+
+/// Check if the graph is weakly connected
+///
+/// :param PyDiGraph graph: The graph to check if it is weakly connected
+///
+/// :returns: Whether the graph is weakly connected or not
+/// :rtype: bool
+///
+/// :raises NullGraph: If an empty graph is passed in
+#[pyfunction]
+#[text_signature = "(graph, /)"]
+pub fn is_weakly_connected(graph: &digraph::PyDiGraph) -> PyResult<bool> {
+    if graph.graph.node_count() == 0 {
+        return Err(NullGraph::new_err("Invalid operation on a NullGraph"));
+    }
+    Ok(weakly_connected_components(graph)[0].len() == graph.graph.node_count())
```

Comment:
Is this line equivalent with `Ok(number_weakly_connected_components(graph) == 1)`?

Reply:
It might be, I'm not sure if there is a case where you can have >1 component and still have the graph fully connected. I based this check on the networkx equivalent function: https://github.com/networkx/networkx/blob/master/networkx/algorithms/components/weakly_connected.py#L144